### PR TITLE
FL Quant

### PR DESCRIPTION
**Pull request by hguermaz**
## Description

* Use FP quantization method if RDOQ ON and INTRA block and 8BIT (x86 WIP)
* Fix txb_skip_context and dc_sign_context update for 128x128 (bug fix).
* Use the existing loop filter multi-mode signal to control loop filter and removed the NRF checks @ Encode Pass and Loop Filter Processes  (bug fix).

## Author

@hguermaz
@JackZhouVCD

## Type of change

New feature/tool and bug fix.

## Tests and performance

* Bd-Rate: 0.6% Bd-Rate gain to aom-cpu0-2pass on the full sub-4K list (the SVT-to-aom-cpu0-2pass BdRate difference dropped from TBD to TBD).
* Speed: 15% slower for M0 on the 1080p aom clips (because of C @ Fp_Quant). Fp_Quant is by default Off untill adding x86 support.  
* Memory footprint: ~0% memory foot print increase.

### CODE REVIEW

File: Source/Lib/Common/ASM_AVX2/highbd_quantize_intrin_avx2.c [hunk 1]
```diff
@@ -54,6 +54,260 @@ static INLINE void init_qp(const int16_t *zbin_ptr, const int16_t *round_ptr,
...
+        round = _mm_add_epi16(round, rnd);
+        round = _mm_srai_epi16(round, log_scale);
+    }
+
+    init_one_qp(&round, &qp[0]);
+    init_one_qp(&quant, &qp[1]);
+
+    if (log_scale == 1) {
+        qp[1] = _mm256_slli_epi16(qp[1], log_scale);
+    }
```
Comment (joelsole) [inline Source/Lib/Common/ASM_AVX2/highbd_quantize_intrin_avx2.c]: no braces

File: Source/Lib/Common/ASM_AVX2/highbd_quantize_intrin_avx2.c [hunk 2]
```diff
@@ -54,6 +54,260 @@ static INLINE void init_qp(const int16_t *zbin_ptr, const int16_t *round_ptr,
...
+static INLINE void read_coeff(const TranLow *coeff, __m256i *c) {
+    if (sizeof(TranLow) == 4) {
+        const __m256i x0 = _mm256_loadu_si256((const __m256i *)coeff);
+        const __m256i x1 = _mm256_loadu_si256((const __m256i *)coeff + 1);
+        *c = _mm256_packs_epi32(x0, x1);
+        *c = _mm256_permute4x64_epi64(*c, 0xD8);
+    }
+    else {
+        *c = _mm256_loadu_si256((const __m256i *)coeff);
+    }
```
Comment (joelsole) [inline Source/Lib/Common/ASM_AVX2/highbd_quantize_intrin_avx2.c]: no braces

File: Source/Lib/Common/Codec/EbFullLoop.c [hunk 3]
```diff
@@ -477,6 +444,276 @@ void av1_quantize_b_facade_II(
...
+    const QmVal *qm_ptr = qparam->qmatrix;
+    const QmVal *iqm_ptr = qparam->iqmatrix;
+
+#if 1
+    quantize_fp_helper_c(coeff_ptr, n_coeffs, p->zbin_QTX, p->round_fp_QTX,
+        p->quant_fp_QTX, p->quant_shift_QTX, qcoeff_ptr,
+        dqcoeff_ptr, p->dequant_QTX, eob_ptr, sc->scan,
+        sc->iscan, qm_ptr, iqm_ptr, qparam->log_scale);
+#else
+    if (qm_ptr != NULL && iqm_ptr != NULL) {
```
Comment (joelsole) [inline Source/Lib/Common/Codec/EbFullLoop.c]: remove #if 1 and the unused code

File: Source/Lib/Common/Codec/EbFullLoop.c [hunk 4]
```diff
@@ -1874,6 +2134,7 @@ void av1_quantize_inv_quantize(
     if (md_context->trellis_quant_coeff_optimization && *eob != 0 && is_inter && component_type == COMPONENT_LUMA) {
 #endif
 #endif
+#endif
```
Comment (joelsole) [inline Source/Lib/Common/Codec/EbFullLoop.c]: It is getting difficult to review the code with so many #if. Could some of them be removed?

File: Source/Lib/Common/Codec/EbFullLoop.c (hunk 3 above)
Comment (hguermaz) [reply Source/Lib/Common/Codec/EbFullLoop.c]: Done. (completely removed the x86 path until matching the C implementation).

File: Source/Lib/Common/Codec/EbFullLoop.c (hunk 4 above)
Comment (hguermaz) [reply Source/Lib/Common/Codec/EbFullLoop.c]: Done. Removed useless RDOQ macros.

File: Source/Lib/Common/ASM_AVX2/highbd_quantize_intrin_avx2.c [hunk 5]
```diff
@@ -54,6 +54,260 @@ static INLINE void init_qp(const int16_t *zbin_ptr, const int16_t *round_ptr,
...
+    return INT16_MAX - _mm_extract_epi16(eob_s, 0);
+}
+
+static INLINE void write_zero(TranLow *qcoeff) {
+    const __m256i zero = _mm256_setzero_si256();
+    if (sizeof(TranLow) == 4) {
+        _mm256_storeu_si256((__m256i *)qcoeff, zero);
+        _mm256_storeu_si256((__m256i *)qcoeff + 1, zero);
+    }
+    else {
```
[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Common/ASM_AVX2/highbd_quantize_intrin_avx2.c]: Done.

**Comment by hassount** [comment]
This kernel should be added to the list: av1_quantize_fp_facade in #262

**Pull request merged**